packages/connection-router-api/src/primitives.rs

```diff
@@ -0,0 +1,187 @@
+use crate::error::*;
+use axelar_wasm_std::nonempty;
+use cosmwasm_schema::cw_serde;
+use cosmwasm_schema::serde::{Deserialize, Serialize};
+use error_stack::{Report, ResultExt};
+use flagset::flags;
+use regex::Regex;
+use schemars::JsonSchema;
+use std::fmt;
+use std::fmt::{Display, Formatter};
+use std::hash::Hash;
+use std::ops::Deref;
+use std::str::FromStr;
+
+#[cw_serde]
+pub struct Message {
+    /// This field can hold arbitrary data. It has only two requirements:
+    /// 1. it must be possible to use its content to find the corresponding message on the source chain
+    /// 2. the ID must uniquely identify the message on the source chain, i.e. no two messages can have the same ID, and no single message can have multiple valid IDs.
+    ///
+    /// IMPORTANT: Verifier contracts must enforce these requirements.
+    pub id: String,
+    pub source_chain: ChainName,
+    pub source_address: Address,
+    pub destination_chain: ChainName,
+    pub destination_address: Address,
+    /// hash length is enforced to be 32 bytes
+    pub payload_hash: [u8; 32],
+}
+
+// [cw_serde] has been expanded here because we need to implement PartialEq manually
+#[derive(
+    ::cosmwasm_schema::serde::Serialize,
+    ::cosmwasm_schema::serde::Deserialize,
+    Clone,
+    Debug,
+    ::cosmwasm_schema::schemars::JsonSchema,
+)]
+#[serde(deny_unknown_fields, crate = "::cosmwasm_schema::serde")]
+#[schemars(crate = "::cosmwasm_schema::schemars")]
+#[serde(try_from = "String")]
+pub struct ChainName(String);
+
+impl Hash for ChainName {
+    /// this is implemented manually because we want to ignore case when hashing
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.to_lowercase().hash(state)
+    }
+}
+
+impl PartialEq for ChainName {
+    /// this is implemented manually because we want to ignore case when checking equality
+    fn eq(&self, other: &Self) -> bool {
+        self.0.to_lowercase() == other.0.to_lowercase()
+    }
+}
+
+impl FromStr for ChainName {
+    type Err = Error;
+
+    fn from_str(chain_name: &str) -> Result<Self, Self::Err> {
+        fn matches_chain_name_pattern(s: &str) -> bool {
```

Comment:
It's a bit weird to define this function. It's not even used for more than once.

Reply:
it just seems natural to implement from_str if we also implement `TryFrom<String>` and it makes tests nicer to write

Reply:
Hmm I meant this local function `matches_chain_name_pattern`.